rust/src/tree.rs

```diff
@@ -1,185 +1,1264 @@
-                    ids: input_ids[curr_idx..].to_vec(),
-                    count: 0,
-                };
-
-                let first_id = new_child.ids[0];
-                curr.children.insert(first_id, new_child);
-
-                curr = curr.children.get_mut(&first_id).unwrap();
-                curr.count += 1;
-                curr_idx = input_ids_len;
             }
         }
     }
 
-    pub fn prefix_match<'a>(&self, input_ids: &'a [u32]) -> &'a [u32] {
-        let mut curr = &self.root;
-
+    pub fn prefix_match(&self, text: &str) -> (String, String) {
+        let mut curr = Arc::clone(&self.root);
         let mut curr_idx = 0;
-        let input_ids_len = input_ids.len();
 
-        while curr_idx < input_ids_len {
-            match curr.children.get(&input_ids[curr_idx]) {
-                Some(child) => {
-                    let prefix_len = common_prefix_len(&input_ids[curr_idx..], &child.ids);
+        let mut prev = Arc::clone(&self.root);
+        let text_count = text.chars().count();
+
+        while curr_idx < text_count {
+            let first_char = text.chars().nth(curr_idx).unwrap();
+            let curr_text = slice_by_chars(text, curr_idx, text_count);
+
+            curr = prev.clone();
 
-                    if prefix_len == child.ids.len() {
-                        curr_idx += prefix_len;
-                        curr = child;
+            match curr.children.entry(first_char) {
+                Entry::Occupied(entry) => {
+                    let matched_node = entry.get().clone();
+                    let shared_count =
+                        shared_prefix_count(&matched_node.text.read().unwrap(), &curr_text);
+
+                    let matched_node_text_count = matched_node.text.read().unwrap().chars().count();
+
+                    if shared_count == matched_node_text_count {
+                        // Full match with current node's text, continue to next node
+                        curr_idx += shared_count;
+                        prev = Arc::clone(&matched_node);
                     } else {
-                        curr_idx += prefix_len;
+                        // Partial match, stop here
+                        curr_idx += shared_count;
+                        prev = Arc::clone(&matched_node);
                         break;
                     }
                 }
-                None => {
+                Entry::Vacant(_) => {
+                    // No match found, stop here
                     break;
                 }
             }
         }
 
-        &input_ids[..curr_idx]
+        curr = prev.clone();
+
+        // Select the first tenant (key in the map)
+        let tenant = curr
+            .tenant_last_access_time
+            .iter()
+            .next()
+            .map(|kv| kv.key().to_owned())
+            .unwrap_or("empty".to_string());
+
+        // Traverse from the curr node to the root and update the timestamp
```

Comment:
Maybe not important, but this could be happened during the matching process (traverse takes time, but maybe is not the bottleneck now).